cls_module/cls_module/memory/stm/aha/msp.py

```diff
@@ -0,0 +1,117 @@
+"""MonosynapticPathway class."""
+
+from collections import defaultdict
+
+import torch
+import torch.nn as nn
+import torch.optim as optim
+import torch.nn.functional as F
+
+from cerenaut_pt_core.utils import build_topk_mask
+from cerenaut_pt_core.components.simple_autoencoder import SimpleAutoencoder
+
+
+class MonosynapticPathway(nn.Module):
+  """
+  An error-driven monosynaptic pathway implementation.
+  """
+
+  def __init__(self, ca3_shape, ec_shape, config):
+    super(MonosynapticPathway, self).__init__()
+
+    self.config = config
+    self.ca1_reset_params = self.config['ca1'].get('reset_params', False)
+    self.ca1_reset_optim = self.config['ca1'].get('reset_optim', False)
+    self.ca3_ca1_reset_params = self.config['ca3_ca1'].get('reset_params', True)
+    self.ca3_ca1_reset_optim = self.config['ca3_ca1'].get('reset_optim', True)
+
+    # Build the CA1 sub-module, to reproduce the EC inputs
+    self.ca1 = SimpleAutoencoder(ec_shape, self.config['ca1'], output_shape=ec_shape)
+    self.ca1_optimizer = optim.Adam(self.ca1.parameters(),
+                                    lr=self.config['ca1']['learning_rate'],
+                                    weight_decay=self.config['ca1']['weight_decay'])
+
+    ca1_output_shape = [1, self.config['ca1']['num_units']]
+
+    # Build the CA1 sub-module, to reproduce the EC inputs
+    self.ca3_ca1 = SimpleAutoencoder(ca3_shape, self.config['ca1'], output_shape=ca1_output_shape)
+    self.ca3_ca1_optimizer = optim.Adam(self.ca3_ca1.parameters(),
+                                        lr=self.config['ca3_ca1']['learning_rate'],
+                                        weight_decay=self.config['ca3_ca1']['weight_decay'])
+
+  def reset(self):
+    if self.ca1_reset_params:
+      self.ca1.reset_parameters()
+
+    if self.ca3_ca1_reset_params:
+      self.ca3_ca1.reset_parameters()
+
+    # Reset the module optimizer
+    if self.ca1_reset_optim:
+      self.ca1_optimizer.state = defaultdict(dict)
```

Comment:
`dict` is a built-in Python keyword to define dicts. It is essentially setting `optimizer.state` to an empty dict i.e. `{}`.

I used this instead as by default PyTorch initialises states like this, so it made sense to just replicate that when resetting  in case something internal is expecting a `defaultdict` object for some reason.